### ISSUE TITLE
Nested Size Fields

### ISSUE THREAD

**Issue by AndyFug** [construct/construct#970]
Hi,
This is a question rather than an issue.

I'm working with a binary protocol (over IP) and this library appears to be exactly what I need.  I admit that I'm new to this world, so apologies if the question seems obvious.

A simplified representation of the struct I need to work with is this:
```python
message = Struct(
    "header" / Struct(
        "version" / Const(2, Int8ub),
        "header_length" / Int8ub,  # Size of header
        "message_length" / Int8ub,  # Size of entire message
        "misc_field" / Prefixed(Int8ub, CString("utf_8"))
    ),
    "payload" / Struct(
        "value" / Const(5, Int16ub)  # Could be many different things, but using this as placeholder
    )
)
```
My question is how should I work with the `header_length` and `message_length` fields?  I've tried using `Tell` and `Rawcopy`.  My best effort is below (which doesn't feel like the correct solution).  This appears to build correctly, but doesn't parse back to the header container correctly (showing None):
```python
header = Struct(
    "version" / Const(2, Int8ub),
    "header_length" / Padding(1),
    "message_length" / Padding(1),
    "misc_field" / Prefixed(Int8ub, CString("utf_8"))
)

payload = Struct(
    "value" / Const(5, Int16ub)
)

message = Struct(
    "header" / RawCopy(header),
    "header_length" / Pointer(this.header.offset1 + Padding(1).sizeof(), Rebuild(Int8ub, this.header.length)),
    "payload" / RawCopy(payload),
    "message_length" / Pointer(this.header.offset1 + 2*Padding(1).sizeof(), Rebuild(Int8ub, this.payload.offset2))
)

built = message.build(dict(
    header=dict(value=dict(
        misc_field="Hello World"
    ))
))

print(built)
print(message.parse(built))
```
Running the above gives:
```
b'\x02\x10\x12\x0cHello World\x00\x00\x05'
Container: 
    header = Container: 
        data = b'\x02\x10\x12\x0cHello World\x00' (total 16)
        value = Container: 
            version = 2
            header_length = None
            message_length = None
            misc_field = u'Hello World' (total 11)
        offset1 = 0
        offset2 = 16
        length = 16
    header_length = 16
    payload = Container: 
        data = b'\x00\x05' (total 2)
        value = Container: 
            value = 5
        offset1 = 16
        offset2 = 18
        length = 2
    message_length = 18

Process finished with exit code 0

```
Any help and suggestions on how best to approach this would be greatly appreciated.

Many thanks in advance.

Andy

**Comment by AndyFug** [comment]
Hi,

Thanks for responding.  
### HiQnet Protocol
The documentation for the protocol I'm trying to parse and build is here:

[HiQnet 3rd Party Programmer Documentation](https://adn.harmanpro.com/site_elements/resources/515_1414083576/HiQnet_Third_Party_Programmers_Guide_v2_original.pdf)
[HiQnet 3rd Party Quick Start Guide](https://adn.harmanpro.com/site_elements/resources/692_1414084052/HiQnet_third-party_programmers_quick-start_guide_original.pdf)

Please note that, as I mention in my original post, my example is a simplified representation, so you're going to find in the spec above that the field sizes will be different etc.  However, you should see that the header length and message length are both encoded in fields within the header rather than prefixed, which is what I'm struggling to implement properly.
### Padding(1)
In response to your remarks about the Padding fields.  I have read the docs several times, but couldn't find (or understand) the tools in the libary that I needed to use to build/parse those length fields.  I tried using `Tell` and `RawCopy` with `Pointer` to achieve what I needed and I came up with the use of `Padding(1)` by looking at this example you put on [this page ](https://construct.readthedocs.io/en/latest/tunneling.html#compression-and-checksuming)of the docs where you inserted a parsed value (in this case, a checksum) into a previous field:

```python
import hashlib
d = Struct(
    "offset" / Tell,
    "checksum" / Padding(64),
    "fields" / RawCopy(Struct(
        Padding(1000),
    )),
    "checksum" / Pointer(this.offset, Checksum(Bytes(64),
        lambda data: hashlib.sha512(data).digest(),
        this.fields.data)),
)
d.build(dict(fields=dict(value={})))
```
Here you use `Padding(64)` seemingly as a placeholder for a value you will build later, so that's why I put `Padding(1)` into my code.  I may well have completely misunderstood what was going on here and not applied it to my use-case correctly at all.

I agree that using Padding(1).sizeof() was silly... I was trying not to hardcode '1' into that part of the code and wanted to use something like `this.header_length.sizeof()` to gain my relative offset dynamically.  I don't think that worked, so I started playing around and left that Padding(1) behind.  I should have simplified that in my example to you.
### String Scheme
Regarding the mixing and matching of two string schemes... yes... I agree.  It seemed odd to me as well (even as a novice).  I read and re-read section 3.5.1 (Page 41 of 91) of the [3rd Party Programmer Documentation](https://adn.harmanpro.com/site_elements/resources/515_1414083576/HiQnet_Third_Party_Programmers_Guide_v2_original.pdf) from above, and it clearly talks about a NULL terminated string and pre-fixing a 'UWORD' with the string length including the NULL character.  I played around with both `PascalString` and `CString` and using Adaptors etc, but landed on what you see above.  Unlike in my example above, I also think I need to use `"utf_16_be"` encoding for the string and `Int16ub` for the prefixed length, but this wasn't part of my question so I just put in some simpler fields to help communicate my question.  If there's a simpler implementation of the spec than my example above, let me know.

I hope that clears up at least some of the reasoning behind my 'contraption'.  Apologies again if any of this is bizzarre or obvious to you - I'm just trying to have a go at implementing something, and this is my first stab at building/parsing structs.

Regards,
Andy

**Comment by AndyFug** [comment]
Thanks for your response.

> We are talking about 3 HiQnet Message Format from page 22 right?

Yes.  Correct.  That page shows the header structure.  Note that it is variable in length due to the optional 'error header', which has a string in it.  The equivalent to this in my example at the start is the `"misc_field"` for simplicity.

> You could use this little experimental invention. That way you can decouple the length field from the Prefixed instance and keep other stuff sandwitched in between. Does this help?

Forgive me, but I'm not quite sure how this helps me with calculating the "headerlength" when building from a `dict()`.  I tried your suggestion, but when building this I have to pass in the header length like `dict(headerlength=3, header=dict(...))`, which my higher level application code won't know, especially since the header struct is of variable length.  The other issue I see is how you'd calculate the payload_length field?

I'll try to re-present the example.  See my comments in-line below:
```python
message = Struct(
    "version" / Const(2, Int8ub),  # Always 2
    "header_length" / Byte,  # To be calculated automatically when building and parsed to a Container when parsing
    "payload_length" / Byte,  # To be calculated automatically when building and parsed to a Container when parsing
    "header" / Struct(...),  # This struct might have a variable length field such as an 'error message' string
    "payload" / Struct(...)  # This struct might have a variable length field such as 'parameter name' string or whatever
)

# The message struct should allow me to build using something like this
a = message.build(dict(header=dict(...), payload=dict(...)))
print(a)
```

1. Is it possible to compute the `"header_length"` and `"payload_length"` fields automatically when building from a dict? 
2. Is it possible with the same Struct to parse back in the opposite direction and be left with a Container that has the `"header_length"` and `"payload_length"` fields populated?

**Comment by arekbulski** [comment]
You misrepresented one field. In the spec there is no payload length, its (whole) message length. 

I think you should do this: Compute the two inner structs separately. Move the header struct and payload struct outside. Build them. Then insert those bytes into Message struct, with two Bytes fields instead. The length fields can then be Rebuild that compute simple formulas over the bytes objects. 

Does this make sense?

**Comment by AndyFug** [comment]
> You misrepresented one field. In the spec there is no payload length, its (whole) message length.

You're quite right.   My original example was correct but I admit I rushed that last post on the train home. I understand that "message_length" as I should have called it needs to compute the whole thing... That's one of the tricky things I am trying to overcome.

I think I understand what you're proposing I do. I'll go away and try to reproduce that and report back.  Many thanks again for your time.

**Comment by arekbulski** [comment]
Another misatake is that the Version field should be part of the Header as well... Still thinking how to build this.

**Comment by AndyFug** [comment]
> Another mistake is that the Version field (and both lengths) should be part of the Header as well... Still thinking how to build this.

Yea I was aware of that.  I imagined that we'd extract the "version" field and not include it as part of the "header" field.  I assume it would be possible to add the size of the version field onto the values for the size of the header and header+payload.

**Comment by AndyFug** [comment]
So I had another go taking on board your advice.  See here:

```python
header = Struct(
    "msg_id" / Int8ub,
    "error_msg" / Prefixed(Int16ub, CString("utf_16_be"))
)

payload = Struct(
    "value" / Const(2, Int8ub)
)

message = Struct(
    "version" / Const(2, Int8ub),
    "header_length" / Int8ub,
    "message_length" / Int8ub,
    "header" / Bytes(lambda this: this.header_length - this._subcons.version.sizeof()),
    "payload" / Bytes(this.message_length - this.header_length)
)

# Build header & payload separately
_header_built = header.build(dict(msg_id=11, error_msg="Error message"))
_payload_built = payload.build(dict())

# Calculate header & payload lengths, taking into account the version field
header_length = len(_header_built) + 1  # +1 for "version" field
message_length = header_length + len(_payload_built)

_message = message.build(dict(header_length=header_length,
                              message_length=message_length,
                              header=_header_built,
                              payload=_payload_built))

print(f"Built/encoded representation:\n{_message.hex(':')}")

parsed_message = message.parse(_message)
parsed_header = header.parse(parsed_message.header)
parsed_payload = payload.parse(parsed_message.payload)

print("\nParsed message:\n", parsed_message)
print("\nParsed header:\n", parsed_header)
print("\nParsed payload:\n", parsed_payload)

```
```
Built/encoded representation:
02:20:21:0b:00:1c:00:45:00:72:00:72:00:6f:00:72:00:20:00:6d:00:65:00:73:00:73:00:61:00:67:00:65:00:00:02

Parsed message:
 Container: 
    version = 2
    header_length = 32
    message_length = 33
    header = b'\x0b\x00\x1c\x00E\x00r\x00r\x00o\x00r\x00 \x00'... (truncated, total 31)
    payload = b'\x02' (total 1)

Parsed header:
 Container: 
    msg_id = 11
    error_msg = u'Error message' (total 13)

Parsed payload:
 Container: 
    value = 2

Process finished with exit code 0

```
I think this looks ok, and I can see it working within my application.  (NB: I know some of the fields above are not strictly in line with the spec, but I've simplified it for the purposes of working through the question in hand).

The only issue I think I still have with it is that I'd prefer the "header_length" & "message_length" calculation to take place within the "message" struct so when calling `message.build(...)` I don't have to pass in those lengths.  Where I'm doing `len(_header_built) + 1` outside the struct, it assumes more intimate knowledge of the struct it's about to be inserted into (that there is a version field that should be included as part of the header length).  Clearly the struct isn't going to change, so it's totally workable, but I'd be interested to know if there was a better approach that would mean not having to pass in the length fields when building.  Maybe it just means using an `Adapter` ??

**Comment by arekbulski** [comment]
As I said, you could replace the length fields with a Rebuild instead of Byte. However, that will just move the formulas from the script to the declarations. But if you want that, why not.

```py
    "header" / Bytes(lambda this: this.header_length - this._subcons.version.sizeof()),

header_length = len(_header_built) + 1  # +1 for "version" field
```
Those should be -3 and +3 because the two length fields are part of the header.

Btw, when reading the specs I keep seeing this "STRING" type all over it but I cannot find a definition for it. What the fuck is it supposed to be? A pascal string, a c string, utf8 or utf16 or what?

**Comment by AndyFug** [comment]
> Those should be -3 and +3 because the two length fields are part of the header.

Quite right.

> As I said, you could replace the length fields with a Rebuild instead of Byte. However, that will just move the formulas from the script to the declarations. But if you want that, why not.

Apologies, but I'm not quite sure how I make it work with `Rebuild` because I think I end up with a circular dependency.  Ie... I need to set the Bytes length of the 'header' field, but I assume I need to make that dependent on the 'header_length' field.
```python
message = Struct(
    "version" / Const(2, Int8ub),
    "header_length" / Rebuild(Int8ub, lambda this: this._subcons.header.sizeof() + 3),
    "message_length" / Rebuild(Int8ub, lambda this: this.header_length + this._subcons.payload.sizeof()),
    "header" / Bytes(lambda this: this.header_length - 3),
    "payload" / Bytes(this.message_length - this.header_length)
)
```
Running the above raises exceptions (I won't show the full trace unless you need it):
```
line 56, in <lambda>
    "header" / Bytes(lambda this: this.header_length - 3),
  File "...\lib\site-packages\construct\lib\containers.py", line 93, in __getattr__
    raise AttributeError(name)
KeyError: 'header_length'

During handling of the above exception, another exception occurred:

File "...\lib\site-packages\construct\lib\containers.py", line 93, in __getattr__
    raise AttributeError(name)
AttributeError: header_length

During handling of the above exception, another exception occurred:

  File "...\lib\site-packages\construct\core.py", line 861, in _sizeof
    raise SizeofError("cannot calculate size, key not found in context", path=path)
construct.core.SizeofError: Error in path (sizeof) -> header
cannot calculate size, key not found in context
```

> Btw, when reading the specs I keep seeing this "STRING" type all over it but I cannot find a definition for it. What the fuck is it supposed to be? A pascal string, a c string, utf8 or utf16 or what?

As I mentioned in my 2nd post above, see **Section 3.5.1 (Page 41 of 91)** of the [3rd Party Programmer Documentation](https://adn.harmanpro.com/site_elements/resources/515_1414083576/HiQnet_Third_Party_Programmers_Guide_v2_original.pdf) (See the bottom of page 41 and top of page 42 - pgs 42/43 of the PDF)  It took me a while to find that information and I thought even as a novice that it was a little odd to mix those two schemes .

**Comment by arekbulski** [comment]
My my, what is it with that obsession to use sizeof everywhere? You are using Rebuild wrong:
```py
"header_length" / Rebuild(Int8ub, len_(this.header) + 3,
"message_length" / Rebuild(Int8ub, this.header_length + len_(this.payload) ),
```
The specs are shit. It uses the term "Unicode" as if it was synonymous with "UTF16BE" which is bollocks.

**Comment by AndyFug** [comment]
Ah... of course.  Because 'header' field is now a Bytes object in the context we can just measure that rather than using the subcon.  Thank you for your help and your way with words. :)  Definitely got a better approach to this now.

> The specs are shit. It uses the term "Unicode" as if it was synonymous with "UTF16BE" which is bollocks.

Indeed.  Doesn't help when I'm trying to find my way with this sort of stuff in general... I think UTF16BE is what they mean, but I guess I'll find out when I try to talk to a device.

**Comment by arekbulski** [comment]
> because I think I end up with a circular dependency.

I think you do not quite understand how Struct parsing and building work. If you look at Struct source, in the build method, the subcon is serialized into the stream and only at the end its ret value is inserted into the context dictionary. Got it? Now look at your code. When header_length is built, the lambda is evaluated, and only afterwards the context key will be inserted. So the lambda uses this somewhat magical key "_subcons" to get the header Bytes field and calls its sizeof method. Fine. So what does the sizeof do? All it does is do a context dictionary lookup for a "header_length" key. That key was not inserted yet so the lookup fails, thus causing a SizeofError. 

The only two ways that I can imagine that could cause a circular evaluation would be either use circular sizeofs (which you dont do) or use circular Lazy classes (which you are not near of).

**Comment by AndyFug** [comment]
> I think you do not quite understand how Struct parsing and building work. 

More than likely. :)

> That key was not inserted yet so the lookup fails, thus causing a SizeofError.

Indeed.  Maybe it's my incorrect use of terminology (almost certainly), but since the "header" Bytes field is dependent on the value of the "header_length" context entry (which isn't created yet), which in turn is dependent on the size of the "header" field/Struct, doesn't that make it a circular dependency (or reference)?  The "header_length" field can't build until it knows the size of the "header" Struct and the "header" Struct can't be fully declared until it knows the context entry for "header_length".

Of course I may still not understand the building/parsing process, and you're quite right, I should spend some more time examining the source.

> The specs are shit. It uses the term "Unicode" as if it was synonymous with "UTF16BE" which is bollocks.

BTW, do you think the field I've declared is what the spec requires?  (however bizarre it may seem) :

`    "error_msg" / Prefixed(Int16ub, CString("utf_16_be"))`

**Comment by arekbulski** [comment]
> doesn't that make it a circular dependency (or reference)? 

```py
    "header_length" / Rebuild(Int8ub, lambda this: this._subcons.header.sizeof() + 3),
    "header" / Bytes(lambda this: this.header_length - 3),
```
No. The "header" lambda only looks up a context key of name "header_length" and does not refer to the "header_length" field. You are mixing the field with the context entry. They share the name but are not the same thing.

> BTW, do you think the field I've declared is what the spec requires?

Yes, with two details. One, the specs do not say integers are big endian so Int16ub is a hypothesis. Secondly, this field is optional so you should add the flags field and use an IF class.

**Comment by arekbulski** [comment]
If this is not a mindfuck enough then consider the following construction. Dont lookup the answer but think about it and tell me, is this circular or not?
```py
Struct(
    "A" / Rebuild(Byte, lambda this: this._subcons.B.sizeof() ),
    "B" / Bytes(lambda this: this._subcons.A.sizeof() ),
)
```
The correct answer is that it is not circular. Consider building the Struct. Struct builds top to bottom so it starts with A.build. Rebuild.build calls the value lambda to obtain the value that will be passed to Byte.build. The value lambda calls B.sizeof. Bytes.sizeof calls the length lambda, which calls A.sizeof. Then... Rebuild.sizeof does NOT call the value lambda but rather it calls the Byte.sizeof which is by definition 1. This ends it for building A. Building B is partially the same.

**Comment by AndyFug** [comment]
Hi,

Sorry for the slow response.

> Yes, with two details. One, the specs do not say integers are big endian so Int16ub is a hypothesis. Secondly, this field is optional so you should add the flags field and use an IF class.

I agree that in the main documentation it doesn't talk about the endianness of the numeric fields, but if you look at the [Quick Start Documentation](https://adn.harmanpro.com/site_elements/resources/692_1414084052/HiQnet_third-party_programmers_quick-start_guide_original.pdf) it states in the first line that numeric fields are all sent over the wire in 'network byte order' or 'big-endian'.  In the main documentation, it does talk about big-endian in Section 6 under RS232 Packet Service, but doesn't seem to mention it elsewhere.

> The correct answer is that it is not circular. Consider building the Struct. Struct builds top to bottom so it starts with A.build. Rebuild.build calls the value lambda to obtain the value that will be passed to Byte.build. The value lambda calls B.sizeof. Bytes.sizeof calls the length lambda, which calls A.sizeof. Then... Rebuild.sizeof does NOT call the value lambda but rather it calls the Byte.sizeof which is by definition 1. This ends it for building A. Building B is partially the same.

I get it... I did what you said and thought it through... I didn't quite see it through to the correct answer, but now I've read yours I think I can see where I went wrong.  I think I'm getting a better understanding of where I mixed up fields and context entries and how the wrappers fit in with that mix during build/parse operations.

I've been implementing what we've spoken about in my application... just tricky finding the time to do it at the moment.  I'll share some snippets to show where this ended up.

On a side note, I've had a glance at [construct_typed](https://pypi.org/project/construct-typing/), and am interested in the `DataclassStruct`, particularly to help with code analysis and type hints etc in my IDE.  I need to look at this a little further (to check there's not gotchas for my use-case), but it sounds like it may help.
